src/iceberg/expression/inclusive_metrics_evaluator.cc

```diff
@@ -0,0 +1,544 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "iceberg/expression/inclusive_metrics_evaluator.h"
+
+#include "iceberg/expression/binder.h"
+#include "iceberg/expression/expression_visitor.h"
+#include "iceberg/expression/rewrite_not.h"
+#include "iceberg/manifest_entry.h"
+#include "iceberg/schema.h"
+#include "iceberg/transform.h"
+#include "iceberg/util/macros.h"
+
+namespace iceberg {
+
+namespace {
+static const bool ROWS_MIGHT_MATCH = true;
+static const bool ROWS_CANNOT_MATCH = false;
+static const int32_t IN_PREDICATE_LIMIT = 200;
+}  // namespace
+
+class InclusiveMetricsVisitor : public BoundVisitor<bool> {
+ public:
+  explicit InclusiveMetricsVisitor(const DataFile& data_file) : data_file_(data_file) {}
+
+  Result<bool> AlwaysTrue() override { return ROWS_MIGHT_MATCH; }
+
+  Result<bool> AlwaysFalse() override { return ROWS_CANNOT_MATCH; }
+
+  Result<bool> Not(bool child_result) override { return !child_result; }
+
+  Result<bool> And(bool left_result, bool right_result) override {
+    return left_result && right_result;
+  }
+
+  Result<bool> Or(bool left_result, bool right_result) override {
+    return left_result || right_result;
+  }
+
+  Result<bool> IsNull(const std::shared_ptr<BoundTerm>& term) override {
+    if (IsNonNullPreserving(term)) {
```

Comment:
```suggestion
    // if the column has no null values, the expression cannot match
    if (IsNonNullPreserving(term)) {
```

Let's preserve the original comment